swupd/manifest.go

```diff
@@ -319,3 +342,215 @@ func (m *Manifest) WriteManifestFile(path string) error {
+	sort.Slice(m.DeletedFiles, func(i, j int) bool {
+		if m.DeletedFiles[i].Version < m.DeletedFiles[j].Version {
+			return true
+		}
+		if m.DeletedFiles[i].Version > m.DeletedFiles[j].Version {
+			return false
+		}
+		return m.DeletedFiles[i].Name < m.DeletedFiles[j].Name
+	})
+}
+
+func (m *Manifest) addDeleted(oldManifest *Manifest) {
+	for _, df := range oldManifest.DeletedFiles {
+		if df.findFileNameInSlice(m.Files) == nil {
+			m.Files = append(m.Files, df)
+			m.DeletedFiles = append(m.DeletedFiles, df)
+		}
+	}
+}
+
+// linkPeersAndChange
+// At this point Manifest m should only have the files that were present
+// in the chroot for that manifest. Link delta peers with the oldManifest
+// if the file in the oldManifest is not deleted or ghosted
+func (m *Manifest) linkPeersAndChange(oldManifest *Manifest) int {
+	changed := 0
+	for _, of := range oldManifest.Files {
+		if match := of.findFileNameInSlice(m.Files); match != nil {
+			if of.Status == statusDeleted || of.Status == statusGhosted {
+				continue
+			}
+
+			match.DeltaPeer = of
+			of.DeltaPeer = match
+			if !sameFile(match, of) {
+				match.Version = m.Header.Version
+				changed++
+			} else {
+				match.Version = of.Version
+			}
+		}
+	}
+
+	return changed
+}
+
+func (m *Manifest) filesAdded(oldManifest *Manifest) int {
+	added := 0
+	for _, af := range m.Files {
+		if af.findFileNameInSlice(oldManifest.Files) == nil {
+			af.Version = m.Header.Version
+			added++
+		}
+	}
+
+	return added
+}
+
+func (m *Manifest) newDeleted(oldManifest *Manifest) int {
+	deleted := 0
+	for _, df := range oldManifest.Files {
+		if df.Status != statusDeleted && df.findFileNameInSlice(m.Files) == nil {
+			if df.Status == statusGhosted {
+				continue
+			}
+			df.Version = m.Header.Version
+			df.Status = statusDeleted
+			df.Modifier = modifierUnset
+			df.Type = typeUnset
+			m.Files = append(m.Files, df)
+			m.DeletedFiles = append(m.DeletedFiles, df)
+			deleted++
+		}
+	}
+
+	return deleted
+}
+
+func (m *Manifest) readIncludes(bundles []*Manifest, c config) error {
+	// read in <imageBase>/<version>/noship/<bundle>-includes
+	var err error
+	path := filepath.Join(c.imageBase, fmt.Sprint(m.Header.Version), "noship", m.Name+"-includes")
+	bundleNames, err := readIncludesFile(path)
+	if err != nil {
+		return err
+	}
+
+	includes := []*Manifest{}
+	for _, b := range bundles {
+		if b.Name == "os-core" {
+			includes = append(includes, b)
+		}
+	}
+
+	for _, bn := range bundleNames {
+		for _, b := range bundles {
+			if bn == b.Name {
+				includes = append(includes, b)
+			}
+		}
+	}
+
+	m.Header.Includes = includes
+	return nil
+}
+
+func compareIncludes(m1 *Manifest, m2 *Manifest) bool {
+	return reflect.DeepEqual(m1.Header.Includes, m2.Header.Includes)
+}
+
+func (m *Manifest) hasTypeChanges() bool {
+	for _, f := range m.Files {
+		if f.typeHasChanged() {
+			return true
+		}
+	}
+
+	return false
+}
+
+func (m *Manifest) subtractManifestFromManifest(m2 *Manifest) {
```

Comment:
Optional: since it contains some non-trivial logic, this function is a good candidate for having tests.